### PR TITLE
Apex code for code review

### PR DESCRIPTION
**Pull request by ghost**
Apex code for code review

### CODE REVIEW

File: Permission and Access Helper/force-app/main/default/classes/ComboboxDataWrapper.cls [hunk 1]
```diff
@@ -0,0 +1,12 @@
+public with sharing class ComboboxDataWrapper {
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/ComboboxDataWrapper.cls:1]: Add copyright comment for a file.

Comment (rupesh-p1) [reply Permission and Access Helper/force-app/main/default/classes/ComboboxDataWrapper.cls:1]: Also for other files

File: Permission and Access Helper/force-app/main/default/classes/FieldPermissionsSelector.cls [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        return FieldPermissions.getSObjectType();
+    }
+    public static IFieldPermissionSelector newInstance()
+    {
+        return (IFieldPermissionSelector) PermissionApplication.Selector.newInstance(FieldPermissions.getSObjectType());
+    }
+    public List<FieldPermissions> getAllRecords(){
+        return (List<FieldPermissions>) Database.query(
+            newQueryFactory()
+            .selectFields(new List<String>{'Parent.ProfileId', 'Parent.IsOwnedByProfile'})
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/FieldPermissionsSelector.cls:23]: we can generalize this method by taking fields as a argument

File: Permission and Access Helper/force-app/main/default/classes/FieldPermissionsSelector.cls [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+            .setLimit(5)
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/FieldPermissionsSelector.cls:24]: why we need to set limit here ?

File: Permission and Access Helper/force-app/main/default/classes/FieldPermissionsSelector.cls-meta.xml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>57.0</apiVersion>
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/FieldPermissionsSelector.cls-meta.xml:3]: Sync Api version for all the files. There are some files with 56 apiVersion and some files with 57 apiVersion

File: Permission and Access Helper/force-app/main/default/classes/IPermissionService.cls [hunk 5]
```diff
@@ -0,0 +1,8 @@
+public interface IPermissionService {
+    List<ComboboxDataWrapper> getProfile();
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/IPermissionService.cls:2]: getProfile -> getProfiles

File: Permission and Access Helper/force-app/main/default/classes/IPermissionService.cls [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+    List<ComboboxDataWrapper> getPermissionSet();
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/IPermissionService.cls:3]: getPermissionSet -> getPermissionSets

File: Permission and Access Helper/force-app/main/default/classes/ObjectPermissionSelector.cls [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    public static IObjectPermissionSelector newInstance()
+    {
+        return (IObjectPermissionSelector) PermissionApplication.Selector.newInstance(ObjectPermissions.getSObjectType());
+    }
+
+    public List<ObjectPermissions> getAllRecords(){
+        return (List<ObjectPermissions>) Database.query(
+            newQueryFactory()
+            .selectFields(new List<String>{'Parent.ProfileId', 'Parent.IsOwnedByProfile'})
+            .setLimit(5)
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/ObjectPermissionSelector.cls:29]: why limit ?

File: Permission and Access Helper/force-app/main/default/classes/ObjectPermissionSelector.cls [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    public static IObjectPermissionSelector newInstance()
+    {
+        return (IObjectPermissionSelector) PermissionApplication.Selector.newInstance(ObjectPermissions.getSObjectType());
+    }
+
+    public List<ObjectPermissions> getAllRecords(){
+        return (List<ObjectPermissions>) Database.query(
+            newQueryFactory()
+            .selectFields(new List<String>{'Parent.ProfileId', 'Parent.IsOwnedByProfile'})
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/ObjectPermissionSelector.cls:28]: generalize this method by taking fields as argument

File: Permission and Access Helper/force-app/main/default/classes/PermissionController.cls [hunk 9]
```diff
@@ -0,0 +1,33 @@
+public class PermissionController {
+    
+    @AuraEnabled(cacheable=true)
+    public static List<ComboboxDataWrapper> getProfile(){
+        return PermissionService.getProfile();
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/PermissionController.cls:5]: put  PermissionService.getProfile(); in try catch and throw AuraHandledException  if any exception occurs

Comment (rupesh-p1) [reply Permission and Access Helper/force-app/main/default/classes/PermissionController.cls:5]: same for all controller methods

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls [hunk 10]
```diff
@@ -0,0 +1,264 @@
+public with sharing class PermissionServiceImpl implements IPermissionService{
+    
+ 	public List<ComboboxDataWrapper> getProfile(){
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:3]: format this file by prettier or manually

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+        Integer keyFortype = 0;
+        Integer keyForPermissionProfile = 0;
+        Integer keyForPermissionType = 0;     
+        Integer keyForPermission = 0;   
+        List<PermissionDataWrapper> wrappedData = new List<PermissionDataWrapper>();
+        Map<String, List<ObjectPermissions>> objectPermissionByProfile= new Map<String, List<ObjectPermissions>>();
+        Map<String, List<ObjectPermissions>> objectPermissionByPermissionSet = new Map<String, List<ObjectPermissions>>();
+        System.debug(objectPermissions);
+        for(ObjectPermissions opData : objectPermissions){
+            System.debug(opData.Parent.IsOwnedByProfile);
```
Comment (rupesh-p1) [inline Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:95]: Remove unnecessary system debug

File: Permission and Access Helper/force-app/main/default/classes/ObjectPermissionSelector.cls (hunk 7 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/ObjectPermissionSelector.cls:29]: Limits were put in place for testing purposes, same will be removed in the coming changes

File: Permission and Access Helper/force-app/main/default/classes/PermissionControllerTest.cls [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+
+        //Then
+        ((IPermissionService)mocks.verify(serviceMock)).getPermissionSet();
+        System.assertEquals(forCheck , recievedPermissionList);
+
+    }
+
+    @isTest
+    private static void  test_getAllPermissionData_shouldReturnAllPermissionData(){
```
Comment (meta-jitendra-singh) [inline Permission and Access Helper/force-app/main/default/classes/PermissionControllerTest.cls:68]: the test method name should in the pattern of given_when_then

File: Permission and Access Helper/force-app/main/default/classes/PermissionControllerTest.cls [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        fflib_ApexMocks mocks = new fflib_ApexMocks();
+        IPermissionService serviceMock = (IPermissionService)mocks.mock(PermissionServiceImpl.class);
+
```
Comment (meta-jitendra-singh) [inline Permission and Access Helper/force-app/main/default/classes/PermissionControllerTest.cls:71]: Also add what is given , which condition is in when and then that you assert

File: Permission and Access Helper/force-app/main/default/classes/PermissionControllerTest.cls-meta.xml [hunk 14]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>55.0</apiVersion>
```
Comment (meta-jitendra-singh) [inline Permission and Access Helper/force-app/main/default/classes/PermissionControllerTest.cls-meta.xml:3]: Please consistent apiVersion in all xml

File: Permission and Access Helper/force-app/main/default/classes/PermissionDataWrapper.cls [hunk 15]
```diff
@@ -0,0 +1,58 @@
+public with sharing class PermissionDataWrapper {
+
+    @AuraEnabled
+    public String type;
+    @AuraEnabled
+    public Integer key;
+    @AuraEnabled
+    public List<PermissionSetProfileData> permissionSetProfileData;
+    @AuraEnabled
+    public Integer rowspan;
```
Comment (meta-jitendra-singh) [inline Permission and Access Helper/force-app/main/default/classes/PermissionDataWrapper.cls:10]: what was the use of rowspan?

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+    }
+    public List<ComboboxDataWrapper> getObjectOptionsForObjectPermissions(){
+        List<ObjectPermissions> objectPermissions = ObjectPermissionSelector.newInstance().getAllRecords();
+        List<ComboboxDataWrapper> wrappedData = new List<ComboboxDataWrapper>();
+        Set<String> objectsInUse = new Set<String>();
+        for(ObjectPermissions op : objectPermissions){
+            objectsInUse.add(op.SobjectType);
+        }
+        List<String> objectApinameList = new List<String>();
+        objectApinameList.addAll(objectsInUse);
```
Comment (meta-jitendra-singh) [inline Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:25]: why we create new list here , we can use the ObjectsInUse list

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+        List<ObjectPermissions> objectPermissions = ObjectPermissionSelector.newInstance().getAllRecords();
+        List<PermissionSet> permissionSets = PermissionSetSelector.newInstance().getAllRecords();
+        List<Profile> profiles = ProfileSelector.newInstance().getAllRecords();
+        List<PermissionDataWrapper> dataToReturn = convertAllDataToWrapper(fieldPermissions, objectPermissions, permissionSets, profiles);
+        return dataToReturn;
+    }
+    public List<ComboboxDataWrapper> getObjectOptionsForObjectPermissions(){
+        List<ObjectPermissions> objectPermissions = ObjectPermissionSelector.newInstance().getAllRecords();
+        List<ComboboxDataWrapper> wrappedData = new List<ComboboxDataWrapper>();
+        Set<String> objectsInUse = new Set<String>();
```
Comment (meta-jitendra-singh) [inline Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:20]: use meaningful name for variables

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+        objectApinameList.addAll(objectsInUse);
+        objectApinameList.sort();
+        Map<String, SObjectType> objectMap = Schema.getGlobalDescribe();
+        for(String obj : objectApinameList){
+            SObjectType objectFromMap = objectMap.get(obj);
+            String label = obj;
+            if(objectFromMap != null){
+                label = objectFromMap.getDescribe().getLabel();
+            }
+            wrappedData.add(new ComboboxDataWrapper(label, obj));
```
Comment (meta-jitendra-singh) [inline Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:55]: what was the difference when the objectFromMap is null then we create wrapper with same value why?

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+    public List<ComboboxDataWrapper> getFieldOptionsByObject(String objectName){
+        List<FieldPermissions> fieldPermissions = FieldPermissionsSelector.newInstance().getRecordsByObject(objectName);
+        List<ComboboxDataWrapper> wrappedData = new List<ComboboxDataWrapper>();
+        List<Schema.DescribeSObjectResult> descResult = Schema.describeSObjects(new String[]{objectName});
+        Map<String, Schema.SObjectField> fields = descResult[0].fields.getMap();
+        Set<String> fieldsInFP = new Set<String>();
+        for(FieldPermissions fp : fieldPermissions){
+            fieldsInFP.add(fp.Field.split('\\.')[1]);
+        }
+        System.debug(fieldsInFP);
```
Comment (meta-jitendra-singh) [inline Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:68]: please get rid of debug

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls [hunk 20]
```diff
@@ -0,0 +1,264 @@
...
+                    objectPermissionByPermissionSet.put(opData.ParentId, nameList);
+                }else{
+                    objectPermissionByPermissionSet.put(opData.ParentId, new List<ObjectPermissions> {opData});
+                }
+            }
+        }
+        Map<String, List<FieldPermissions>> fieldPermissionByProfile= new Map<String, List<FieldPermissions>>();
+        Map<String, List<FieldPermissions>> fieldPermissionByPermissionSet = new Map<String, List<FieldPermissions>>(); 
+        for(FieldPermissions fpData : fieldPermissions){
+            if(fpData.Parent.IsOwnedByProfile){
```
Comment (meta-jitendra-singh) [inline Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:117]: can we get rid from multiple if else?

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls [hunk 21]
```diff
@@ -0,0 +1,264 @@
...
+                }else{
+                    fieldPermissionByPermissionSet.put(fpData.ParentId, new List<FieldPermissions> {fpData});
+                }
+            }
+        }
+        if(objectPermissionByProfile.size() != 0 || fieldPermissionByProfile.size() != 0){
+            PermissionDataWrapper pdw = new PermissionDataWrapper();
+            pdw.type = 'Profile';
+            pdw.key = keyFortype++;
+            Integer pdwRowspan = 0;
```
Comment (meta-jitendra-singh) [inline Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:139]: this is for what?

**Review by meta-jitendra-singh — CHANGES_REQUESTED** [review]

File: Permission and Access Helper/force-app/main/default/classes/IPermissionService.cls (hunk 5 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/IPermissionService.cls:2]: Done

File: Permission and Access Helper/force-app/main/default/classes/IPermissionService.cls (hunk 6 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/IPermissionService.cls:3]: Done

File: Permission and Access Helper/force-app/main/default/classes/PermissionController.cls (hunk 9 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/PermissionController.cls:5]: Done

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls (hunk 10 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:3]: Done

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls (hunk 11 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:95]: Done

File: Permission and Access Helper/force-app/main/default/classes/FieldPermissionsSelector.cls (hunk 3 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/FieldPermissionsSelector.cls:24]: Done

File: Permission and Access Helper/force-app/main/default/classes/IPermissionService.cls (hunk 5 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/IPermissionService.cls:2]: Done

File: Permission and Access Helper/force-app/main/default/classes/FieldPermissionsSelector.cls-meta.xml (hunk 4 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/FieldPermissionsSelector.cls-meta.xml:3]: Done

File: Permission and Access Helper/force-app/main/default/classes/PermissionControllerTest.cls-meta.xml (hunk 14 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/PermissionControllerTest.cls-meta.xml:3]: Done Sir

File: Permission and Access Helper/force-app/main/default/classes/PermissionDataWrapper.cls (hunk 15 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/PermissionDataWrapper.cls:10]: Rowspan was used for the attribute rowspan in the actual table, where we needed to set the rowspan for merging rows

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls (hunk 17 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:20]: Can you please specify as to what variable you are referring to as all the variables in the visible code snippet have been named meaningfully already

File: Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls (hunk 18 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/PermissionServiceImpl.cls:55]: There are some cases in which the object is not fetched in getGlobalDescribe, which were discussed earlier, so for them, we have to add this condition such that if the object is not fetched, it's API name is used as it's label

File: Permission and Access Helper/force-app/main/default/classes/ComboboxDataWrapper.cls (hunk 1 above)
Comment (metavikas) [reply Permission and Access Helper/force-app/main/default/classes/ComboboxDataWrapper.cls:1]: Should it be done just for the Apex classes or the JS and HTMLs too ?

File: Permission and Access Helper/force-app/main/default/classes/PermissionDataWrapper.cls (hunk 15 above)
Comment (meta-jitendra-singh) [reply Permission and Access Helper/force-app/main/default/classes/PermissionDataWrapper.cls:10]: I think we should handle this at client side

**Pull request merged**